cscs-checks/microbenchmarks/latency/src/latency.cpp

```diff
@@ -0,0 +1,230 @@
+
+        buffer_size = (buffer_size / PAGESIZE + 1) * PAGESIZE;
+
+        unsigned group_size;
+        if (simulate_large_pages) {
+            unsigned npages = buffer_size / PAGESIZE;
+            // use 48 out of 64 entries in TLB-L1
+            unsigned ngroups = npages / 48 + 1;
+            unsigned pages_per_group = npages / ngroups;
+            group_size = pages_per_group * PAGESIZE;
+            buffer_size = group_size * ngroups;
+        }
+        else
+            group_size = buffer_size;
+
+        unsigned num_elements = buffer_size / sizeof(CacheLine);
+        unsigned group_elements = group_size / sizeof(CacheLine);
+
+        size_t reps = estimate_reps(buffer_size, L1, L2, L3);
+        if (argc > 2) reps = std::stoi(argv[2]);
+
+        CacheLine* chain;
+        if (posix_memalign(reinterpret_cast<void**>(&chain),
+                           PAGESIZE, num_elements * sizeof(CacheLine)))
+            throw std::bad_alloc();
+
+        std::vector<unsigned> path = random_grouped_path(num_elements, group_elements);
+        convert_to_pointers(chain, path);
+
+        double test = time_function(chainload, chain, reps/10);
+        reps = reps * 0.25/(test*10);
+
+        double duration = 0;
+        duration += time_function(chainload, chain, reps);
+
+        double ns_per_load = duration / reps * 1e9;
+        double clocks = GHZ * ns_per_load;
+
+        std::cout << "latency (ns): " << ns_per_load
+                  << " clocks: " << clocks << " size " << buffer_size
+                  << " ngroups: " << buffer_size/group_size << std::endl;
+
+        std::free(chain);
+    }
+}
+
+
+/**********************************************/
+
+
+void convert_to_pointers(CacheLine* start, std::vector<unsigned> const& path)
+{
+    for (unsigned i = 0; i < path.size(); ++i)
+        start[i].payload = start + path[i];
+}
+
+
+size_t chainload(CacheLine* next, size_t reps)
+{
+    // compiler doesn't unroll
+    for (size_t i = 0; i < reps; i+=4)
+    {
+        next = next->payload;
+        next = next->payload;
+        next = next->payload;
+        next = next->payload;
+    }
+
+    return reinterpret_cast<size_t>(next);
+}
+
+
+// create a random permutation of numbers [0,length-1]
+std::vector<unsigned> random_permutation(unsigned length)
+{
+    static unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    static auto gen = std::default_random_engine(seed);
+
+    std::vector<unsigned> ret(length);
+    unsigned var = 0;
+
+    std::generate(ret.begin(), ret.end(), [&var]() { return var++; } );
+    std::shuffle(ret.begin(), ret.end(), gen);
+
+    return ret;
+}
+
+
+// convert any permutation visit_sequence to a path
+// --
+// A path is equivalent to a permutation of [0, length-1]
+// that contains no valid permutation in any subset
+// [0,n] for n < length-1
+// --
+// e.g for length=5, 12034 would be invalid, because 120 is
+// a valid permutation of [0,2], leading to a closed loop
```

Comment:
This is correct, but I can't see how do you check for that? How do you guarantee that you can't have such a permutation?

Reply:
The way I construct the path from the permutation guarantees that all elements are visited.
To see this, one needs to realize that a random permutation \sigma interpreted as a sequence of elements to be visited is not the same as the array of indices (the path) that traverses the elements by dependency chaining in the order specified by \sigma.

The function make_path creates the dependency chain for the input permutation \sigma.
Since \sigma is a valid permutation, it contains all elements in the range, and therefore the path traverses all elements.

e.g. the example given: 12034 is a valid permutation (but not a valid path)
the corresponding path (visit [1], then [2], [0], [3], [4] and then again [1]) is 32041.
This will produce the order 12034 in the for loop

`int next = sigma[0];`
`for (int i = 0; i < sigma.size(); ++i)`
`    next = path[next];`

Reply:
I got it, you are right. It's a bit confusing the comment as it is now. The mathematical definition is correct, but I would like to see in the comment how you achieve that. I guess that the simply stating clearly the difference between the `visit_sequence` and `path` would help + the example of how the `12034` permutation gets into `32041` (not step-by-step of course, just to mention it). So sth like the following would be enough:

- `visit_sequence` is the order that we want to visit the nodes
- Each element of the `path` contains a "pointer" (index) to the next element to visit according to the visit sequence. For example, a `visit_sequence` of 12034 will produce as path `32041`.